src/crypto/pwhash/argon2id13.rs

```diff
@@ -0,0 +1,601 @@
+            0xe1, 0xd2, 0x5c, 0x4a, 0xdd, 0x80, 0x8e, 0x21, 0x56, 0x97, 0xad, 0x3f, 0x77, 0x9a,
+            0x51, 0xa4, 0x62, 0xf8, 0xbf, 0x52, 0x61, 0x0a, 0xf2, 0x1f, 0xc6, 0x9d, 0xba, 0x6b,
+            0x07, 0x26, 0x06, 0xf2, 0xda, 0xbc, 0xa7, 0xd4, 0xae, 0x1d, 0x91, 0xd9, 0x19,
+        ];
+        let result = run_derive_key_test(&password, salt, &expected, 167, 3, 1784128);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_derive_09() {
+        let password = [
+            0xa3, 0x47, 0xae, 0x92, 0xbc, 0xe9, 0xf8, 0x0f, 0x6f, 0x59, 0x5a, 0x44, 0x80, 0xfc,
+            0x9c, 0x2f, 0xe7, 0xe7, 0xd7, 0x14, 0x8d, 0x37, 0x1e, 0x94, 0x87, 0xd7, 0x5f, 0x5c,
+            0x23, 0x00, 0x8f, 0xfa, 0xe0, 0x65, 0x57, 0x7a, 0x92, 0x8f, 0xeb, 0xd9, 0xb1, 0x97,
+            0x3a, 0x5a, 0x95, 0x07, 0x3a, 0xcd, 0xbe, 0xb6, 0xa0, 0x30, 0xcf, 0xc0, 0xd7, 0x9c,
+            0xaa, 0x2d, 0xc5, 0xcd, 0x01, 0x1c, 0xef, 0x02, 0xc0, 0x8d, 0xa2, 0x32, 0xd7, 0x6d,
+            0x52, 0xdf, 0xbc, 0xa3, 0x8c, 0xa8, 0xdc, 0xbd, 0x66, 0x5b, 0x17, 0xd1, 0x66, 0x5f,
+            0x7c, 0xf5, 0xfe, 0x59, 0x77, 0x2e, 0xc9, 0x09, 0x73, 0x3b, 0x24, 0xde, 0x97, 0xd6,
+            0xf5, 0x8d, 0x22, 0x0b, 0x20, 0xc6, 0x0d, 0x7c, 0x07, 0xec, 0x1f, 0xd9, 0x3c, 0x52,
+            0xc3, 0x10, 0x20, 0x30, 0x0c, 0x6c, 0x1f, 0xac, 0xd7, 0x79, 0x37, 0xa5, 0x97, 0xc7,
+            0xa6,
+        ];
+        let salt = [
+            0x55, 0x41, 0xfb, 0xc9, 0x95, 0xd5, 0xc1, 0x97, 0xba, 0x29, 0x03, 0x46, 0xd2, 0xc5,
+            0x59, 0xde,
+        ];
+        let expected = [
+            0x20, 0xe7, 0xba, 0x6f, 0xaa, 0x2c, 0x0a, 0x4b, 0x07, 0xf3, 0xff, 0x38, 0xe1, 0x5e,
+            0x25, 0x2a, 0x06, 0x9c, 0x2c, 0x62, 0xba, 0xc3, 0xf2, 0x78, 0x5d, 0x31, 0x17, 0x64,
+            0xd7, 0x3e, 0x67, 0xfd, 0x71, 0x3b, 0xe3, 0x42, 0xee, 0x93, 0x8e, 0x6d, 0xf4, 0xde,
+            0x6a, 0xf1, 0xa8, 0x9a, 0x44, 0xb8, 0x58, 0x98, 0x38, 0x86, 0x44, 0x57, 0xbc, 0xfe,
+            0x3c, 0xf0, 0xf2, 0xd3, 0x29, 0xb8, 0x00, 0xab, 0x9f, 0x58, 0x10, 0xb6, 0x32, 0x55,
+            0x88, 0xeb, 0x4e, 0x0c, 0x56, 0xf9, 0x91, 0x92, 0xb2, 0xcc, 0x76, 0xdc, 0x81, 0x94,
+            0xdc, 0x10, 0x97, 0xfe, 0x5e, 0xd1, 0x2a, 0xc4, 0x21, 0x44, 0x81, 0xc0, 0x3c, 0x35,
+            0x97, 0x13, 0x1b, 0xa1, 0x64, 0xa5, 0x6e, 0x71, 0x87, 0xe2, 0xda, 0x56, 0x5a, 0x8c,
+            0xd5, 0x29, 0x66, 0x8e, 0x9a, 0x37, 0xfa, 0xa5, 0x8a, 0x17, 0x01, 0xc4, 0x9a, 0x14,
+            0xed, 0xf7, 0xa5, 0x0d, 0xec, 0x41, 0x43, 0xb4, 0x56, 0xcb, 0xa6, 0xd1, 0x4c, 0x95,
+            0x7b, 0xb6, 0x55, 0xe9, 0x9c, 0xe9, 0x6b, 0xc5, 0x06, 0x96, 0x12, 0x16, 0xef, 0x88,
+            0x7a,
+        ];
+        let result = run_derive_key_test(&password, salt, &expected, 155, 4, 397645);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_derive_10() {
+        let password = [
+            0xa3, 0x47, 0xae, 0x92, 0xbc, 0xe9, 0xf8, 0x0f, 0x6f, 0x59, 0x5a, 0x44, 0x80, 0xfc,
```

Comment:
@kvnallsn where did you take these vectors?

Reply:
These are from libsodium itself, from test/default/pwhash_argon2*

Reply:
Initially I had an implementation of hex2bin and bin2hex but removed them in favor of using the raw vectors and wait for #252 to land.  I don't see much reason to change this until that patch gets merged. Even then, I think the raw vectors fit well here since the test methodology is different. Libsodium does all tests then compares the resulting output to a reference file whereas we can compare results individually.